metrics/internal/stackdriver/secret_watcher.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stackdriver
+
+import (
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Observer is the interface for callbacks that notify an Observer of the latest
+// state of a secret being watched by a secretWatcher.  An Observer should not modify the provided
+// Secrets, and should `.DeepCopy()` it for persistence (or otherwise process its
+// contents).
+type Observer interface {
+	OnAdd(*corev1.Secret)
+	OnUpdate(*corev1.Secret, *corev1.Secret)
+	OnDelete(*corev1.Secret)
+}
+
+// ObserverFuncs implements the observer interface.
+type ObserverFuncs struct {
+	AddFunc    func(*corev1.Secret)
+	UpdateFunc func(*corev1.Secret, *corev1.Secret)
+	DeleteFunc func(*corev1.Secret)
+}
+
+// OnAdd is the OnAdd of the observer interface.
+func (of *ObserverFuncs) OnAdd(s *corev1.Secret) {
+	if of.AddFunc == nil {
+		return
+	}
+	of.AddFunc(s)
+}
+
+// OnUpdate is the OnUpdate of the observer interface.
+func (of *ObserverFuncs) OnUpdate(sOld *corev1.Secret, sNew *corev1.Secret) {
+	if of.UpdateFunc == nil {
+		return
+	}
+	of.UpdateFunc(sOld, sNew)
+}
+
+// OnDelete is the OnDelete of the observer interface.
+func (of *ObserverFuncs) OnDelete(s *corev1.Secret) {
+	if of.DeleteFunc == nil {
+		return
+	}
+	of.DeleteFunc(s)
+}
+
+// SecretWatcher defines the interface that a secret watcher must implement.
+type SecretWatcher interface {
+	// StartWatch starts the secret watch allowing observer callbacks to be triggered.
+	StartWatch() error
+
+	// StopWatch stops the secret watch.
+	StopWatch()
+
+	// GetNamespaceWatched returns the namespace of the Secret being watched;
+	// returns empty string if more than one namespace is being watched.
+	GetNamespaceWatched() string
```

Comment:
It is purely for convenience. I imagine it could be useful for logging. For example, something like "Stopped SecretWatcher watching {watcher.GetNamespace()/watcher.GetNameWatched()}".